runtime/lua/vim/lsp/completion.lua

```diff
@@ -0,0 +1,773 @@
+local M = {}
+
+local api = vim.api
+local lsp = vim.lsp
+local protocol = lsp.protocol
+local ms = protocol.Methods
+
+local rtt_ms = 50
+local ns_to_ms = 0.000001
+
+--- @alias vim.lsp.CompletionResult lsp.CompletionList | lsp.CompletionItem[]
+
+-- TODO(mariasolos): Remove this declaration once we figure out a better way to handle
+-- literal/anonymous types (see https://github.com/neovim/neovim/pull/27542/files#r1495259331).
+--- @nodoc
+--- @class lsp.ItemDefaults
+--- @field editRange lsp.Range | { insert: lsp.Range, replace: lsp.Range } | nil
+--- @field insertTextFormat lsp.InsertTextFormat?
+--- @field insertTextMode lsp.InsertTextMode?
+--- @field data any
+
+--- @nodoc
+--- @class vim.lsp.completion.BufHandle
+--- @field clients table<integer, vim.lsp.Client>
+--- @field triggers table<string, vim.lsp.Client[]>
+
+--- @type table<integer, vim.lsp.completion.BufHandle>
+local buf_handles = {}
+
+--- @nodoc
+--- @class vim.lsp.completion.Context
+local Context = {
+  cursor = nil, --- @type { [1]: integer, [2]: integer }?
+  last_request_time = nil, --- @type integer?
+  pending_requests = {}, --- @type function[]
+  isIncomplete = false,
+  suppress_completeDone = false,
+  expand_snippet = false,
```

Comment:
I think it's possible to remove `suppress_completeDone`, `expand_snippet` and the `accept_pum()` function once https://github.com/neovim/neovim/pull/20671 gets merged